### ISSUE TITLE
Concurrent update to HashSet _ids in IODispatcherAsync (Rare)

### ISSUE THREAD

**Issue by Lougarou** [EventStore/EventStore#2015]
**Describe the bug**
A concurrent update to HashSet _ids in IODispatcherAsync's BeginReadForward or BeginReadBackward will throw a _System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct._

If this occurs when projections are being loaded it can leave them in a weird state. Additionally the read operation is not retried so the projections are left in that state and stuck.

**To Reproduce**
1. Since it can be hard to reproduce. Modify code to throw an InvalidOperationException once when projections are being loaded.

**Expected behavior**
1. Should not crash. Either read operation should be retried or if HashSet is changed to something like a ConcurrentDictionary.

**Actual behavior**
Projections are stuck in a weird state. 
![Screen Shot](https://user-images.githubusercontent.com/4313217/66377526-d354fb80-e9c2-11e9-8ffd-56e882e70d17.png)

**EventStore details**
 - EventStore server version: 5.0.2 (affects previous versions also)

**Comment by rastrup** [comment]
We have just experienced this issue also, from event log

`Application: EventStore.ClusterNode.exe
CoreCLR Version: 4.700.20.47201
.NET Core Version: 3.1.9
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
   at EventStore.Core.Helpers.IODispatcher.ReadBackward(String streamId, Int64 fromEventNumber, Int32 maxCount, Boolean resolveLinks, ClaimsPrincipal principal, Action`1 action, Nullable`1 corrId) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Core\Helpers\IODispatcher.cs:line 0
   at EventStore.Projections.Core.Services.Processing.EmittedStream.SubmitListEvents(CheckpointTag upTo, Int64 fromEventNumber) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Projections.Core\Services\Processing\EmittedStream.cs:line 397
   at EventStore.Projections.Core.Services.Processing.EmittedStream.Handle(ReadTimeout message) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Projections.Core\Services\Processing\EmittedStream.cs:line 429
   at EventStore.Core.Services.TimerService.ThreadBasedScheduler.DoTiming() in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Core\Services\TimerService\ThreadBasedScheduler.cs:line 72
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()`

EventStore server version 20.10.0

**Comment by rastrup** [comment]
The exception is a new protection of non-concurrent collections introduced in .Net Core 2.1
https://github.com/dotnet/runtime/issues/26868

**Comment by rastrup** [comment]
We observed the issue again, new Stack Trace
`System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
   at EventStore.Core.Helpers.IODispatcher.WriteEvents(String streamId, Int64 expectedVersion, Event[] events, ClaimsPrincipal principal, Action`1 action) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Helpers\\IODispatcher.cs:line 576
   at EventStore.Projections.Core.Services.Processing.EmittedStreamsWriter.WriteEvents(String streamId, Int64 expectedVersion, Event[] events, ClaimsPrincipal writeAs, Action`1 complete) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStreamsWriter.cs:line 18
   at EventStore.Projections.Core.Services.Processing.EmittedStream.PublishWriteEvents(Int32 retryCount) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 627
   at EventStore.Projections.Core.Services.Processing.EmittedStream.SubmitWriteEvents() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 563
   at EventStore.Projections.Core.Services.Processing.EmittedStream.SubmitWriteEventsInRecoveryLoop(Boolean anyFound) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 717
   at EventStore.Projections.Core.Services.Processing.EmittedStream.OnWriteCompleted() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 649
   at EventStore.Core.Helpers.IODispatcher.<>c__DisplayClass35_0.<WriteEvents>b__0(WriteEventsCompleted res) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Helpers\\IODispatcher.cs:line 590
   at EventStore.Core.Messaging.RequestResponseDispatcher`2.Handle(TResponse message) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Messaging\\RequestResponseDispatcher.cs:line 69
   at EventStore.Core.Messaging.RequestResponseDispatcher`2.EventStore.Core.Bus.IHandle<TResponse>.Handle(TResponse message) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Messaging\\RequestResponseDispatcher.cs:line 48
   at EventStore.Core.Bus.MessageHandler`1.TryHandle(Message message) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Bus\\MessageHandler.cs:line 30
   at EventStore.Core.Bus.InMemoryBus.Publish(Message message) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Bus\\InMemoryBus.cs:line 290
   at EventStore.Core.Bus.QueuedHandlerMRES.ReadFromQueue(Object o) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Bus\\QueuedHandlerMRES.cs:line 119`

EventStore server version 20.10.0

**Comment by oletolshave** [comment]
We have seen more or less this exact same issue, running on EventStore 20.10.2 in production under load. The result of the exception was the entire event store process crashed and had to be restarted.

It seems the stack trace is more the same as mentioned by @rastrup above, except the exception message is now different (perhaps due to running on .NET 5 instead of .NET core 3.1?).

Running on Windows 2016 with 16 GB of ram.

Stack trace as reported from Windows Event viewer:
```
Application: EventStore.ClusterNode.exe
CoreCLR Version: 5.0.421.11614
.NET Version: 5.0.4
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value, Int32& location)
   at EventStore.Core.Helpers.IODispatcher.ReadBackward(String streamId, Int64 fromEventNumber, Int32 maxCount, Boolean resolveLinks, ClaimsPrincipal principal, Action`1 action, Nullable`1 corrId) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Core\Helpers\IODispatcher.cs:line 0
   at EventStore.Projections.Core.Services.Processing.EmittedStream.SubmitListEvents(CheckpointTag upTo, Int64 fromEventNumber) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Projections.Core\Services\Processing\EmittedStream.cs:line 397
   at EventStore.Projections.Core.Services.Processing.EmittedStream.Handle(ReadTimeout message) in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Projections.Core\Services\Processing\EmittedStream.cs:line 424
   at EventStore.Core.Services.TimerService.ThreadBasedScheduler.DoTiming() in D:\a\TrainStation\TrainStation\build\oss-eventstore\src\EventStore.Core\Services\TimerService\ThreadBasedScheduler.cs:line 75
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

```

Around 2 minutes before the crash, we had these messages repeated in the EventStore log many times:
```
{"@t":"2021-07-27T14:20:44.4323644Z","@mt":"Failed to write events to stream {stream}. Error: {e}","stream":"$ce-EventStoreMessageProcessor","e":"WrongExpectedVersion","SourceContext":"EventStore.Projections.Core.Services.Processing.CoreProjectionCheckpointManager","ProcessId":904,"ThreadId":38}
{"@t":"2021-07-27T14:20:44.4323932Z","@mt":"Projection '{projection}'({projectionCorrelationId}) restart has been requested due to: '{reason}'","projection":"$by_category","projectionCorrelationId":"76b81533-2149-45c4-a1a6-401635fe9d21","reason":"The '$ce-EventStoreMessageProcessor' stream has been written to from the outside. Expected Version: 152857836, Current Version: 152857844. Checkpoint: C:252717851202/P:252717851202.","SourceContext":"EventStore.Projections.Core.Services.Processing.ProcessingStrategySelector","ProcessId":904,"ThreadId":38}
(These lines repeated around 187 times in 2 minutes)

```
and then the final:
```
{"@t":"2021-07-27T14:22:32.8924095Z","@mt":"Global Unhandled Exception occurred.","@l":"Fatal","@x":"System.IndexOutOfRangeException: Index was outside the bounds of the array.\r\n   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value, Int32& location)\r\n   at EventStore.Core.Helpers.IODispatcher.ReadBackward(String streamId, Int64 fromEventNumber, Int32 maxCount, Boolean resolveLinks, ClaimsPrincipal principal, Action`1 action, Nullable`1 corrId) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Helpers\\IODispatcher.cs:line 0\r\n   at EventStore.Projections.Core.Services.Processing.EmittedStream.SubmitListEvents(CheckpointTag upTo, Int64 fromEventNumber) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 397\r\n   at EventStore.Projections.Core.Services.Processing.EmittedStream.Handle(ReadTimeout message) in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Projections.Core\\Services\\Processing\\EmittedStream.cs:line 424\r\n   at EventStore.Core.Services.TimerService.ThreadBasedScheduler.DoTiming() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Services\\TimerService\\ThreadBasedScheduler.cs:line 75\r\n   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)\r\n   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)\r\n--- End of stack trace from previous location ---\r\n   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)\r\n   at System.Threading.ThreadHelper.ThreadStart()","ProcessId":904,"ThreadId":11}
{"@t":"2021-07-27T14:22:36.8103180Z","@mt":"\n{description,-25} {version} ({branch}/{hashtag}, {timestamp})","description":"ES VERSION:","version":"20.10.2.0","branch":"tags/oss-v20.10.2","hashtag":"7c3b13465","timestamp":"Fri, 5 Mar 2021 16:29:41 +0100","SourceContext":"EventStore.Core.EventStoreHostedService`1[[EventStore.Core.ClusterNodeOptions, EventStore.Core, Version=20.0.0.0, Culture=neutral, PublicKeyToken=null]]","ProcessId":23416,"ThreadId":1}
...

```

**Comment by timothycoleman** [comment]
This was recently fixed in multiple branches by #3270 #3271 #3272
